EIPS/eip-3779.md

```diff
@@ -0,0 +1,180 @@
+   var stack[1024]unsigned
+   var SP := 0
+
+   func validate(PC :=0) boolean {
+      // traverse code sequentially
+      // recurse for subroutines and conditional jumps
+      while true {
+         instruction = bytecode[PC]
+         if is_invalid(instruction) {
+            return false;
+         }
+
+         // stack depth must be the same
+         if SP != stack_depth[PC] {
+               return false
+         }
+         // if stack depth non-zero we have been here before 
+         // return true to break cycle in control flow graph
+         if stack_depth[PC] != 0 { 
+            return true
+         }
+         stack_depth[PC] = SP
+
+          if instruction == JUMP {
+
+             // check for valid destination
+             jumpdest = stack[SP]
+             if bytecode[jumpdest] != JUMPDEST {
+                 return false
+             }
+
+             // reset PC to destination of jump 
+             PC += jumpdest
+             continue
+         }
+         if instruction == JUMPI {
+
+             // check for valid destination
+             jumpdest = stack[SP]
+             if bytecode[jumpdest] != JUMPDEST {
+                 return false
+             }
+
+             // reset PC to destination of jump 
+             PC += jumpdest
+
+             // recurse to jump to code to validate 
+             if !validate(stack[SP])) {
+                 return false
+             }
+             continue 
+         }
+         
+        // effect of instruction on stack
+         SP -= remove_items(instruction)
+         SP += add_items(instruction)
+         if SP < 0 || 1024 < SP {
+             return false
+         }
+
+         // successful validation of path
+         if     PC > code_len ||
+                instruction == STOP || 
+                instruction == RETURN ||
+                instruction == SUICIDE {
+             return true
+         }
+
+         // advance PC according to instruction
+         PC = advance_pc(PC, instruction)
+      }    
+   }
+```
+
+## Rationale
+
+The alternative to checking validity at creation time is checking it at runtime.  This hurts performance and is a DoS surface.
+
+## Backwards Compatibility
+
+These changes affect the semantics of existing EVM code – the [EVM Object Format deprecated dynamic use of JUMP and JUMPI.  Contracts that make only static use of JUMP and JUMPI remain valid.  Most contracts to not use dynamic jumps.
+
+## Security Considerations
+
+Safety and amenability to static analysis of valid programs are comparable to [EIP-615](https://eips.ethereum.org/EIPS/eip-615),  without imposing syntactic constraints.  Validity can ensured by following the rules given previously, and programs can be validated with the provided algorithm.  The validation algorithm is simple and bounded by the size of the code, allowing for validation at creation time and preventing DoS attacks.  Compilers can easily follow the rules, and in most cases already do.
```

Comment:
Copyright section is required, and EIP links need to be relative.

Note: As mentioned previously, since this EIP references 615 its progress through the process will be held up by it.  If you think 615 will make it through first then that is fine, but if you think 615 may stall out or be withdrawn I recommend extracting the necessary information from it and making this EIP standalone.

```suggestion
Safety and amenability to static analysis of valid programs are comparable to [EIP-615](./eip-615.md),  without imposing syntactic constraints.  Validity can ensured by following the rules given previously, and programs can be validated with the provided algorithm.  The validation algorithm is simple and bounded by the size of the code, allowing for validation at creation time and preventing DoS attacks.  Compilers can easily follow the rules, and in most cases already do.
 
## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```